lib/Conversion/PolynomialToStandard/PolynomialToStandard.cpp

```diff
@@ -802,6 +803,263 @@ func::FuncOp PolynomialToStandard::buildPolynomialModFunc(FunctionType funcType,
+static SmallVector<APInt> precomputeRoots(APInt root, const APInt &cmod,
+                                          unsigned degree) {
+  APInt baseRoot = root;
+  root = 1;
+  SmallVector<APInt> vals(degree);
+  for (unsigned i = 0; i < degree; i++) {
+    vals[i] = root;
+    root = mulMod(root, baseRoot, cmod);
+  }
+  return vals;
+}
+
+static Value computeReverseBitOrder(ImplicitLocOpBuilder &b,
+                                    RankedTensorType type, Value tensor) {
+  unsigned degree = type.getShape()[0];
+  double degreeLog = std::log2((double)degree);
+  assert(std::floor(degreeLog) == degreeLog &&
+         "expected the degree to be a power of 2");
+
+  unsigned indexBitWidth = (unsigned)degreeLog;
+  auto indicesType =
+      RankedTensorType::get(type.getShape(), IndexType::get(b.getContext()));
+
+  SmallVector<APInt> _indices(degree);
+  for (unsigned index = 0; index < degree; index++) {
+    _indices[index] = APInt(indexBitWidth, index).reverseBits();
+  }
+  auto indices = b.create<arith::ConstantOp>(
+      indicesType, DenseElementsAttr::get(indicesType, _indices));
+
+  SmallVector<utils::IteratorType> iteratorTypes(1,
+                                                 utils::IteratorType::parallel);
+  AffineExpr d0;
+  bindDims(b.getContext(), d0);
+  SmallVector<AffineMap> indexingMaps = {AffineMap::get(1, 0, {d0}),
+                                         AffineMap::get(1, 0, {d0})};
+  auto out = b.create<arith::ConstantOp>(
+      type,
+      DenseElementsAttr::get(type, APInt(type.getElementTypeBitWidth(), 0L)));
+  auto shuffleOp = b.create<linalg::GenericOp>(
+      /*resultTypes=*/TypeRange{type},
+      /*inputs=*/ValueRange{indices.getResult()},
+      /*outputs=*/ValueRange{out.getResult()},
+      /*indexingMaps=*/indexingMaps,
+      /*iteratorTypes=*/iteratorTypes,
+      /*bodyBuilder=*/
+      [&](OpBuilder &nestedBuilder, Location nestedLoc, ValueRange args) {
+        ImplicitLocOpBuilder b(nestedLoc, nestedBuilder);
+        auto idx = args[0];
+        auto elem = b.create<tensor::ExtractOp>(tensor, ValueRange{idx});
+        b.create<linalg::YieldOp>(elem.getResult());
+      });
+  return shuffleOp.getResult(0);
+}
+
+struct ConvertNTT : public OpConversionPattern<NTTOp> {
+  ConvertNTT(mlir::MLIRContext *context)
+      : OpConversionPattern<NTTOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      NTTOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+
+    auto polyTy = dyn_cast<PolynomialType>(op.getInput().getType());
+    if (!polyTy) {
+      op.emitError(
+          "Can't directly lower for a tensor of polynomials. "
+          "It is expected that it should be lowered to be done elementwise.");
+      return failure();
+    }
+
+    RingAttr ring = polyTy.getRing();
+    auto inputType = convertPolynomialType(polyTy);
```

Comment:
The converted type should already be present on the `adaptor.getInput().getType()`; the dialect conversion framework uses whatever type converter is provided by this pass to convert the types of all operands and results. The `op` argument to this function contains the original, unconverted operands.

Reply:
Two lines below, I think this means you would also not need the `to_tensor` op at all.